hazelcast-client/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java

```diff
@@ -0,0 +1,144 @@
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.config.ClientUserCodeDeploymentConfig;
+import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.UserCodeDeploymentConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.test.HazelcastParametersRunnerFactory;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.util.FilteringClassLoader;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import org.junit.runners.Parameterized.UseParametersRunnerFactory;
+import usercodedeployment.IncrementingEntryProcessor;
+
+import java.util.Collection;
+import java.util.LinkedList;
+
+import static java.util.Collections.singletonList;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Parameterized.class)
+@UseParametersRunnerFactory(HazelcastParametersRunnerFactory.class)
+@Category({QuickTest.class})
+public class ClientUserCodeDeploymentTest extends HazelcastTestSupport {
+
+    private TestHazelcastFactory factory = new TestHazelcastFactory();
+
+    @After
+    public void tearDown() throws Exception {
+        factory.terminateAll();
+    }
+
+    @Parameterized.Parameter(0)
+    public UserCodeDeploymentConfig.ClassCacheMode classCacheMode;
+    @Parameterized.Parameter(1)
+    public UserCodeDeploymentConfig.ProviderMode providerMode;
+
+
+    @Parameters(name = "ClassCacheMode:{0}, ProviderMode:{1}")
+    public static Collection<Object[]> parameters() {
+        LinkedList<Object[]> parameters = new LinkedList<Object[]>();
+        for (UserCodeDeploymentConfig.ClassCacheMode classCacheMode : UserCodeDeploymentConfig.ClassCacheMode.values()) {
+            for (UserCodeDeploymentConfig.ProviderMode providerMode : UserCodeDeploymentConfig.ProviderMode.values()) {
+                parameters.add(new Object[]{classCacheMode, providerMode});
+            }
+        }
+        return parameters;
+    }
+
+    private Config createNodeConfig() {
+        Config config = new Config();
+        FilteringClassLoader filteringCL = new FilteringClassLoader(singletonList("usercodedeployment"), null);
+        config.setClassLoader(filteringCL);
+        config.getUserCodeDeploymentConfig()
+                .setEnabled(true)
+                .setClassCacheMode(classCacheMode)
+                .setProviderMode(providerMode);
+        return config;
+    }
+
+    private ClientConfig createClientConfig() {
+        ClientConfig config = new ClientConfig();
+        ClientUserCodeDeploymentConfig clientUserCodeDeploymentConfig = new ClientUserCodeDeploymentConfig();
+        clientUserCodeDeploymentConfig.addClass("usercodedeployment.IncrementingEntryProcessor");
```

Comment:
Can you also add tests for addJar API? I see them exist in the ClientUserCodeDeploymentConfigTest but using here also will test the end-to-end feature.